File: 70.climbing-stairs.c
<<<<<<< HEAD
/*
 * @lc app=leetcode id=70 lang=c
 *
 * [70] Climbing Stairs
 */

// @lc code=start
// By DP
//等同f(n) = f(n-1)+f(n-2)
int climbStairs(int n){
    int i; 
    int p = 0, q = 0, r = 1;
    for(i=0;i<n;i++) {
        p = q;
        q = r;
        r = p+q;
    }
    return r;
}

// @lc code=end

=======
/*
 * @lc app=leetcode id=70 lang=c
 *
 * [70] Climbing Stairs
 */

// @lc code=start
// By DP
//等同f(n) = f(n-1)+f(n-2)
int climbStairs(int n){
    int i; 
    int p = 0, q = 0, r = 1;
    for(i=0;i<n;i++) {
        p = q;
        q = r;
        r = p+q;
    }
    return r;
}

// @lc code=end

>>>>>>> 6861f1229a47360993e49170b9b1be7c1dd4f215

File: 70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#
#By DP, time: O(n), space: O(1)
#基本想法n階只有兩種可能:差一階時跨1步和差兩階時跨2步
#所以等同f(n) = f(n-1)+f(n-2)
#假設n = 5, 答案等同到3階的方法數+到4階的方法數
class Solution:
    def climbStairs(self, n: int) -> int:
        f_prevprev = 0
        f_perv = 0
        f_n = 1
        for i in range(n):
            f_prevprev = f_perv
            f_perv = f_n
            f_n = f_prevprev+f_perv
        return f_n

# @lc code=end

